concepts/generics/about.md

```diff
@@ -0,0 +1,77 @@
+# About
+
+# Introduction
+
+Generics are a language feature that enable code reuse.
+For example, it should not matter what type a list holds in order to compute its length.
+That is why the `length` function uses a generic type parameter `a` in its type signature.
+
+```elm
+length : List a -> Int
+```
+
+## Parameters in types
+
+When a type is generic over another type, we use a lowercase name for that other type, and call it a type parameter.
+Two very common examples are the `List` and `Maybe` types.
+
+```elm
+-- Maybe is generic over the type a
+type Maybe a
+    = Nothing
+    | Just a
+
+someString : Maybe String
+someString = Just "hello"
+
+someInt : Maybe Int
+someInt = Just 42
+```
+
+A type can be generic over multiple other types, and thus have multiple type parameters.
+In that case, every independant type gets a different lowercase name.
+Being independant does not prevent them from becoming the same type, it just enables them to be different.
+
+```elm
+type alias Pair a b =
```

Comment:
What do you think of showing an example of a sum type on top of the product type?

```elm
type Either a b
  = This a
  | That b

this : Either a b -> Maybe a
this either =
  case either of
    This a -> Just a
    That _ -> Nothing
```

If it's too much, maybe it can be in about.md but not in introduction.md?

Reply:
Saying this is confusing because Result exists is the same as saying Pair is confusing because Tuples or records exist.
This code demonstrate that multiple generic types can belong in the same product type or in different variants of a sum type.
But if you prefer not to use it, no worries ^^

Reply:
I do think Pair is a bad example too actually. I took it from Matthieu's branch where he had started work on this. I think students should get used to reading and writing good and idiomatic Elm code. Familiarity helps with learning. So I'll change the Pair example to something else.

> This code demonstrate that multiple generic types can belong in the same product type or in different variants of a sum type.

I'd like to use a different example than Either, but I don't understand this statement, so I can't propose something different. I don't see a product type in the code, just a sum type / custom type?

I could definitely come up with an example that uses 2

Reply:
What is not idiomatic about `Pair`? It's a record with two generic fields, I think it's a decent example. Of course coming up with something better is fine.

Pair is the example for product type, and Either for the sum type, together they are the basis for all you can build.

Reply:
I've changed Pair, as it isn't idiomatic, and it was interesting, as it turns out the record type alias' rarely use generics, so I've added a comment to state that.

> Pair is the example for product type, and Either for the sum type, together they are the basis for all you can build. Of course again, coming up with something better is fine :)

So we have examples of product types and sum types in about.md now, so I think we are good?

> This code demonstrate that multiple generic types can belong in the same product type or in different variants of a sum type.

And I think we are all good with this now as well, I've just used the Result type instead of Either.

So I think there is nothing more to do here, but let me know if I've misunderstood something, I am finding it hard to follow this thread.

Reply:
> it turns out the record type alias' rarely use generics, so I've added a comment to state that.

Just a remark, that generics in product types are used in every single elm program. And you could easily define a type alias for that record.

```
Browser.sandbox :
    { init : model
    , view : model -> Html msg
    , update : msg -> model -> model
    }
    -> Program () model msg
```

I also personally use them a lot to make parametric views, especially to pass message builders as arguments like

```
someButton : Config msg -> Html msg
someButton msgConfig = generateButtonWith msgConfig
```

This pattern for views is sometimes called the config pattern, as in Evan's sortable table example https://package.elm-lang.org/packages/billstclair/elm-sortable-table/latest/Table#view

That being said, I don't mind at all if you prefer going another route.

Reply:
I remember reading about the config pattern and looking at the code for the sortable table. I remember it being hard to work out what was going on, but interesting. I can't remember the details now though. I think if we want to teach that one of us needs to write a blog post or something, to make it a bit more approachable. I found it tricky to work out what the pattern was, and I've read and written a lot of elm, so I think it would be too much for a beginner.

It's true that the main program for all elm apps is generic, but people very rarely create a type alias for it, and you rarely have to interact with it. If you look at the code for some elm apps on git hub, or at the source code for common elm packages, it is very easy to find a lot of examples of generic type parameters for custom types, and really hard to find examples of generic type parameters for product types / records, especially ones that you could use as an understandable example.

I'm happy to take this statement out, but I think it is true, having looked at a lot of code trying to find a good example!

Reply:
Having `Maybe` and `Result` in the core elm package definitely tips the scales to more sum types with generics. However, understanding that generics can be used with both building blocks of elm types (sum and product) is important as a learning lesson. And being able to understand these when reading the types in the package doc or in code is almost as important as actually using them ourselves. So I would not devalue usage of generics for product types in the explanations.

As for the config pattern, it’s actually very simple. It’s just fancy name to say "pass behavior as input argument". In OOP, that’s sometimes called "dependency injection" which is another fancy name. It provides great flexibility when you don’t want to encode in your own function all the possible ways it could be used. You just state that some of the behavior/type has to be provided in the function arguments, in a way that you can use inside your function.

Reply:
Hi Matthieu. about.md does have an example of using generics with records / product types, I've just changed it from Pair (which is code that you would never see in real life), to Event (which is code you might write in real life).

I think it's important to show real practical code examples. If a student sees the Pair code, they might think that it is useful for some problem they are solving and use it, when probably they should be using a Tuple or something instead. Seeing a lot of idiomatic code also means that you get a feel for what id idiomatic in Elm.

Anyway, thanks for the simple explanation for the config pattern.

And is there a change you want making here? If so let's discuss ...